File: ScoringBoxes.py
class ScoringBox:
    """
    Base class for Scoring Boxes.
    """
    def __init__(self, name: str):
        self.name = name

    def score_dice(self, dice: list) -> int:
        """Each scoring box expects a list of int not SixSidedDie
        :param dice: an integer list of dice for the turn
        """
        raise NotImplementedError


class NumScoringBox(ScoringBox):
    """Scoring box class for Ones to Sixes
    """
    def __init__(self, name: str, num: int):
        ScoringBox.__init__(self, name)
        self.num = num

    def score_dice(self, dice: list) -> int:
        for die in dice:
            if die == self.num:
                return self.num * dice.count(die)

        return 0


class XOfAKindScoringBox(ScoringBox):
    """Scoring box for 'X' of a kind
    """
    def __init__(self, name: str, of_a_kind: int):
        ScoringBox.__init__(self, name)
        self.of_a_kind = of_a_kind

    def score_dice(self, dice: list) -> int:
        for die in dice:
            if dice.count(die) >= self.of_a_kind:
                return sum(dice)

        return 0


class ThreeOfAKindScoringBox(XOfAKindScoringBox):
    def __init__(self, name):
        XOfAKindScoringBox.__init__(self, name, 3)


class FourOfAKindScoringBox(XOfAKindScoringBox):
    def __init__(self, name):
        XOfAKindScoringBox.__init__(self, name, 4)


class FullHouseScoringBox(ScoringBox):
    def __init__(self, name):
        ScoringBox.__init__(self, name)

    def score_dice(self, dice: list) -> int:
        first = dice[0]
        num_of_first = dice.count(first)

        if num_of_first == 2 or num_of_first == 3:
            num_of_other = 2
            if num_of_first == 2:
                num_of_other = 3
            for die in dice:
                if die != first and dice.count(die) == num_of_other:
                    return 25

        return 0


class SmallStraightScoringBox(ScoringBox):
    def __init__(self, name):
        ScoringBox.__init__(self, name)
        
    def score_dice(self, dice: list) -> int:
        unique_dice = list(set(sorted(dice)))

        if len(unique_dice) < 4:
            return 0

        for i in range(3):
            if unique_dice[i+1] - unique_dice[i] != 1:
                return 0

        return 30


class LargeStraightScoringBox(ScoringBox):
    def __init__(self, name):
        ScoringBox.__init__(self, name)

    def score_dice(self, dice: list) -> int:
        sorted_dice = sorted(dice)
        if sorted_dice == [1, 2, 3, 4, 5] or sorted_dice == [2, 3, 4, 5, 6]:
            return 40

        return 0


class YahtzeeScoringBox(ScoringBox):
    def __init__(self, name):
        ScoringBox.__init__(self, name)

    def score_dice(self, dice: list) -> int:
        if all(dice[0] == d for d in dice):
            return 50

        return 0


class ChanceScoringBox(ScoringBox):
    def __init__(self, name):
        ScoringBox.__init__(self, name)

    def score_dice(self, dice: list) -> int:
        return sum(dice)

File: ScoringBoxes.swift

//
//  ScoringBoxes.swift
//  yahtzee i guess
//
//  Created by William Setterberg on 5/24/16.
//  Copyright © 2016 Joselliam. All rights reserved.
//

import Foundation

class ScoringBox: NSObject, NSCoding {
    
    var name: String
    
    init(name: String) {
        self.name = name
    }
    
    func scoreDice(dice: [Int]) -> Int {
        // same as chance
        return dice.reduce(0, combine: +)
    }
    
    // MARK: NSCoding
    required init(coder aDecoder: NSCoder) {
        name = aDecoder.decodeObjectForKey("name") as! String
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeObject(name, forKey: "name")
    }
}

class NumScoringBox: ScoringBox {
    
    var numToScore: Int
    
    init(name: String, numToScore: Int) {
        self.numToScore = numToScore
        super.init(name: name)
    }
    
    override func scoreDice(dice: [Int]) -> Int {
        return dice.filter({ $0 == self.numToScore }).reduce(0, combine: +)
    }
    
    required init(coder aDecoder: NSCoder) {
        numToScore = aDecoder.decodeIntegerForKey("numToScore")
        super.init(coder: aDecoder)
    }
    
    override func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeInteger(numToScore, forKey: "numToScore")
        super.encodeWithCoder(aCoder)
    }
}

class NOfAKindScoringBox: ScoringBox {
    
    var ofAKind: Int
    
    init(name: String, ofAKind: Int) {
        self.ofAKind = ofAKind
        super.init(name: name)
    }
    
    override func scoreDice(dice: [Int]) -> Int {
        for die in dice {
            if dice.filter({ $0 == die }).count >= self.ofAKind {
                return dice.reduce(0, combine: +)
            }
        }
        
        return 0
    }
    
    required init(coder aDecoder: NSCoder) {
        ofAKind = aDecoder.decodeIntegerForKey("ofAKind")
        super.init(coder: aDecoder)
    }
    
    override func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeInteger(ofAKind, forKey: "ofAKind")
        super.encodeWithCoder(aCoder)
    }
}

class ThreeOfAKindScoringBox: NOfAKindScoringBox {
    init(name: String) {
        super.init(name: name, ofAKind: 3)
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func encodeWithCoder(aCoder: NSCoder) {
        super.encodeWithCoder(aCoder)
    }
}

class FourOfAKindScoringBox: NOfAKindScoringBox {
    init(name: String) {
        super.init(name: name, ofAKind: 4)
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func encodeWithCoder(aCoder: NSCoder) {
        super.encodeWithCoder(aCoder)
    }
}

class FullHouseScoringBox: ScoringBox {
    override func scoreDice(dice: [Int]) -> Int {
        let first = dice[0]
        let numOfFirst = dice.filter({ $0 == first }).count
        
        if numOfFirst == 2 || numOfFirst == 3 {
            let numOfOther = numOfFirst == 3 ? 2 : 3

            for  die in dice {
                if die != first && dice.filter({ $0 == die }).count == numOfOther {
                    return 25
                }
            }
        }
        
        return 0
    }
    
}

class SmallStraightScoringBox: ScoringBox {
    override func scoreDice(dice: [Int]) -> Int {
        let uniqueDice = [Int] (Set(dice)).sort()

        if uniqueDice.count < 4 {
            return 0
        }
        
        // fix for weird dice combination
        if uniqueDice == [1, 3, 4, 5, 6] {
            return 30
        }
        
        for i in 0..<3 {
            if uniqueDice[i+1] - uniqueDice[i] != 1 {
                return 0
            }
        }
        
        return 30
    }
}

class LargeStraightScoringBox: ScoringBox {
    override func scoreDice(dice: [Int]) -> Int {
        let sortedDice = dice.sort()
        if sortedDice == [1, 2, 3, 4, 5] || sortedDice == [2, 3, 4, 5, 6] {
            return 40
        }
        
        return 0
    }
}

class YahtzeeScoringBox: ScoringBox {
    override func scoreDice(dice: [Int]) -> Int {
        if Set(dice).count != 1 {
            return 0
        }
        return 50
    }
}

class ChanceScoringBox: ScoringBox {
    // nothing to do here
}
